### DISCUSSION TITLE
Is there a shortcut to quick toggle/reset view_filter?

### DISCUSSION THREAD

**Discussion by rvalue** [mitmproxy/mitmproxy#6874] [Q&A]
While using mitmproxy, I often have to update the view_filter to get to the requests i am interested in and I am losing too much time  to turn off the view_filter. So what that looks like is, i press the shortcut `f`  & it autocompletes to `set view_filter ''` in the command prompt. then i have entered some regex based filter and hit enter. then i see that there is some error in my regular expression or use of mitmproxy filter expressions. so now, I am deleting the whole view filter expression or adjusting it partially to get to the right thing.

What i am looking for is a quick shortcut like approach or quick way to just delete all the filter expression to look at `@all` the flows and start again with filtering flows.

Any way to do that?

**Comment by mhils** [comment]
This is already a feature request at https://github.com/mitmproxy/mitmproxy/issues/4604. Happy to take a look if someone wants to revive the PR!

**Comment by rvalue** [comment]
hmm, I went through the issue/feature-req and the attached PRs & I would like to talk a bit more on it

_What is a single toggle option for the viewfilter helpful for?_
1. to check if the filter is being applied to 100% of flows they are interested in or if some are left out
2. to not actually toggle the viewfilter but reset/clear the expression because they messed it up

for case 2, this solution is not really helpful. 
@mhils would you be open to accepting solution for both cases?

----
but I think **there is a need to do a major overhaul of the viewfilter** for the way it is probably being used.
this is how most of my flow analysis goes like:
1. I start with a `browser.start` and open the **main website** i am working with
2. I write a filter to filter out all the flows related to ad trackers

now i can start with some real work

3. I remove all the asset related requests to the **main website**
4. I am left with relevant flows to analyze

by this time the view filter expression has grown long enough

5. I realize there are flows to **CDN website** which are dictating the order of requests to the **main website**
a. Now i have to whitelist these requests 
b. **>> Painful edit of the expression that is full of and, or usage <<**
6. I am happy with most the expression rules added

now to study the website behaviour

7.  i add in more rules to the expression
a. let's add a expression rule to just look at json requests
b. let's add a expression rule to just look at javascripts or media assets being sent
8. Now I want to toggle the rules in 7.a & 7.b individually. Flows matching both the rules are relevant, i don't want to keep rewriting the viewfilter again & again, i just want to toggle them as needed

**So basically one needs a way to toggle on/off a part of view filter expression or clear/delete that part of viewfilter expression when not relevant**

@mhils 
1. Is something like this doable in mitmproxy based on how its using `urwid-mitmproxy` & would you allow a PR for this?
If it were not a console based program but UI based, this would be some kind of groups to work with & checkbox to apply what we want. 
2. If we are to assign some keyboard shortcut for something & its already assigned to something else, are you ok to make breaking changes?
3. What is the policy of using unmerged code from the PR #4760  without the individual's approval under the project's MIT license? Can we work off those changes?

**Reply by mhils** [reply]
I'm generally open to mockups on what that should look like, but I also want to make sure we're not overengineering things. Generally speaking, a simple toggle sounds easier to me.

> If we are to assign some keyboard shortcut for something & its already assigned to something else, are you ok to make breaking changes?

We would prefer not to make breaking changes without a good reason. What specifically are you thinking of?

> What is the policy of using unmerged code from the PR https://github.com/mitmproxy/mitmproxy/pull/4760 without the individual's approval under the project's MIT license? Can we work off those changes?

My understanding is that they committed it to their public mitmproxy fork, which is also licensed as MIT. You're fine here.

**Reply by rvalue** [reply]
for clarity, when i say expression rule, this is what i mean

```
:set view_filter '(rule1 & rule 2) | rule3'
```

I am not familiar with urwid or any curses like library but I was looking at the examples at their [website](http://urwid.org/examples/index.html) & just bouncing ideas here,

what if the way to enter the view filter is still via the command prompt but when it comes to toggling the expression rules within the view filter, it would be something like when you press shortcut key `x` for export & it shows a console popup with curl, httpie & other options 

so something similar where, say you press `F` and a popup shows with expression rules within the view filter, one on each line with a way to check like [x] to turn on & off. 

```txt
Flows
<>19:14:54 HTTPS GET  ~googleapis.com /chrome->ariations/seed?osname=mac&channe~ 200 ~ication/x-gzip 56.5k 609ms
  19:14:54 HTTPS POST ~nts.google.com /ListAccounts?gpsia=1&source=ChromiumBrow~ 200 ~plication/json   43b  79ms
  19:15:03 HTTPS POST ~googleapis.com /v1:GetModels?key=AIzaSyBOti4mM-6x9WDnZIj~ 200 ~ion/x-protobuf  1.0k  78ms
  19:15:03 HTTPS GET  ~googleapis.com /downloads?name=1679317318&target=OPTIMIZ~ 200       text/html  259k 799ms
  19:15:03 HTTPS GET  ~googleapis.com /downloads?name=1714403013&target=OPTIMIZ~ 200       text/html  5.0k 204ms
  19:15:04 HTTPS GET  ~googleapis.com /downloads?name=1673999601&target=OPTIMIZ~ 200       text/html  862k 721ms
  19:15:04 HTTPS GET  ~googleapis.com /downloads?name=5&target=OPTIMIZATION_TAR~ 200       text/html  2.1m 1.07s
  19:15:04 HTTPS GET  ~googleapis.c+----------------    FILTER-------------------+00       text/html 44.1k 611ms
  19:15:05 HTTPS GET  ~googleapis.c|                                             |00       text/html  2.6m 865ms
  19:15:05 HTTPS GET  ~googleapis.c|   +-------------------------------+         |00       text/html  4.6k  77ms
  19:15:05 HTTPS GET  ~googleapis.c|   | [x] rule 1                    |         |00       text/html  3.8k 287ms
  19:15:06 HTTPS GET  ~googleapis.c|   | [x] rule 2                    |         |00       text/html  1.4m 668ms
  19:15:54 HTTPS POST ~googleapis.c|   +-------------------------------+         |00 ~plication/json  3.6k 298ms
  19:15:55 HTTP  GET  ~dl.me.gvt1.c|     [x] rule 3                              |00 ~n/octet-stream 34.2k  77ms
  19:15:55 HTTPS POST ~googleapis.c|                                             |00 ~plication/json  170b 278ms
  19:15:55 HTTP  GET  ~dl.me.gvt1.c|                                             |00 ~n/octet-stream  5.1m 698ms
  19:15:56 HTTPS POST ~googleapis.c|                                             |00 ~plication/json  170b 253ms
  19:15:57 HTTP  GET  ~dl.me.gvt1.c|                                             |00 ~n/octet-stream  8.6k  18ms
  19:15:57 HTTPS POST ~googleapis.c|                                             |00 ~plication/json  171b 269ms
  19:15:59 HTTP  GET  ~dl.me.gvt1.c+---------------------------------------------+00 ~n/octet-stream 47.2k 296ms
  19:15:59 HTTPS POST ~googleapis.com /service/update2/json                      200 ~plication/json  171b 318ms
  19:15:59 HTTP  GET  ~dl.me.gvt1.com /edgedl/release2/chrome_component/V3P1l2h~ 200 ~n/octet-stream  5.3k  28ms
  19:15:59 HTTPS POST ~googleapis.com /service/update2/json                      200 ~plication/json  170b 451ms
  19:16:00 HTTP  GET  ~dl.me.gvt1.com /edgedl/release2/chrome_component/adh2zpn~ 200 ~n/octet-stream  6.3k  22ms
   [1/54]                                                                                                   [*:8080]
Flow:    ⏎ Select      D Duplicate   r Replay      x Export      d Delete      m Mark        e Edit
Proxy:   ? Help        q Quit        E Events      O Options     F Filter box  f Filter      w Save flows
                                                                 +-----------+
```

complex viewfilters would need a more complex layout. this would mean we would have to workout a way to represent this appropriately in the command prompt too so the user is aware the expression rule is off but mitmproxy still remembers it. I am not sure what all capabilities the urwid library can provide in here.

with respect to breaking change, i was talking about using `F` shortcut key which is currently being used, would be weird to have it anything else.
